Failure in "test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache"

The parameter `shuffling_queue_capacity` is not used in the unit test. That means it is always `0`.

https://github.com/uber/petastorm/blob/15b35798d6140efe90f8467072dd55d12a8f79c1/petastorm/tests/test_pytorch_dataloader.py#L230

However, if I added `shuffling_queue_capacity` into `extra_loader_params`, test with `shuffling_queue_capacity=20` are all 
 failing.
https://github.com/uber/petastorm/blob/15b35798d6140efe90f8467072dd55d12a8f79c1/petastorm/tests/test_pytorch_dataloader.py#L238

```
(petastorm_venv3.7) root@4d7bc42e93c2:/petastorm/petastorm/tests# pytest -v test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache
==================================================================== test session starts ====================================================================
platform linux -- Python 3.7.9, pytest-6.2.1, py-1.10.0, pluggy-0.13.1 -- /petastorm_venv3.7/bin/python3.7
cachedir: .pytest_cache
rootdir: /petastorm
plugins: forked-1.3.0, timeout-1.4.2, cov-2.11.1, logger-0.5.1
collected 16 items

test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[1-make_batch_reader-20] FAILED                                              [  6%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[1-make_batch_reader-0] PASSED                                               [ 12%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[1-make_reader-20] FAILED                                                    [ 18%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[1-make_reader-0] PASSED                                                     [ 25%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[2-make_batch_reader-20] FAILED                                              [ 31%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[2-make_batch_reader-0] PASSED                                               [ 37%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[2-make_reader-20] FAILED                                                    [ 43%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[2-make_reader-0] PASSED                                                     [ 50%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[3-make_batch_reader-20] FAILED                                              [ 56%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[3-make_batch_reader-0] PASSED                                               [ 62%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[3-make_reader-20] FAILED                                                    [ 68%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[3-make_reader-0] PASSED                                                     [ 75%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[None-make_batch_reader-20] FAILED                                           [ 81%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[None-make_batch_reader-0] PASSED                                            [ 87%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[None-make_reader-20] FAILED                                                 [ 93%]
test_pytorch_dataloader.py::test_batched_data_loader_with_in_memory_cache[None-make_reader-0] PASSED                                                  [100%]
```


Is this a known issue or we just missed that?

@abditag2 @selitvin @tgaddair

Ok, with `pytest -s`, it is showing `_add_many` raises runtime error as below:
```
            retrieved_so_far = None
            for idx in range(5):
>               batch = next(it)

test_pytorch_dataloader.py:257:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../pytorch.py:124: in __iter__
    for batch in self._iter_impl():
../pytorch.py:394: in _iter_impl
    for b in self._iter_impl_worker():
../pytorch.py:441: in _iter_impl_worker
    other_shuffling_buffer.add_many(batch.values())
../reader_impl/pytorch_shuffling_buffer.py:36: in add_many
    return self._add_many(items)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <petastorm.reader_impl.pytorch_shuffling_buffer.BatchedRandomShufflingBuffer object at 0x7f26794ad8d0>
items = [tensor([230000,  50000, 400000, 300000, 390000, 310000, 110000,      0, 180000,
         70000], dtype=torch.int32)]

    def _add_many(self, items):
        if self._done_adding:
            raise RuntimeError('Can not call add_many after done_adding() was called.')

        if not self.can_add():
>           raise RuntimeError('Can not enqueue. Check the return value of "can_enqueue()" to check if more '
                               'items can be added.')
E           RuntimeError: Can not enqueue. Check the return value of "can_enqueue()" to check if more items can be added.

../reader_impl/pytorch_shuffling_buffer.py:238: RuntimeError
--------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------
ERROR    petastorm.pytorch:pytorch.py:128 Iteration on Petastorm DataLoader raise error: RuntimeError('Can not enqueue. Check the return value of "can_enqueue()" to check if more items can be added.')
```

--- a/docs/release-notes.rst
+++ b/docs/release-notes.rst
@@ -6,6 +6,7 @@ Release notes
 
 Release 0.9.9 (unreleased)
 ==========================
+- `PR 643 <https://github.com/uber/petastorm/pull/643>`_: Tests: fix the unit test of batched dataloader with in-memory cache.
 
 
 Release 0.9.8

--- a/petastorm/pytorch.py
+++ b/petastorm/pytorch.py
@@ -291,7 +291,8 @@ class BatchedDataLoader(LoaderBase):
           instance. If set to 0, no shuffling will be done.
         :param inmemory_cache_all: This is an boolean that indicates whether the data should be
             cached in memory or not. When this cache is enabled, expect smaller than requested
-            batch size on the epoch boundaries.
+            batch size on the epoch boundaries. Enabling this cache also requres that shuffling_queue_capacity to be
+            large enough to hold all data.
         :param num_epochs: Number of epochs to load when in memory cache is enabled. If
             set to None, loader will loads batches indefinitely.
         """

--- a/petastorm/tests/test_pytorch_dataloader.py
+++ b/petastorm/tests/test_pytorch_dataloader.py
@@ -227,7 +227,8 @@ def test_mem_cache_num_epochs_without_mem_cache_error(two_columns_non_petastorm_
             BatchedDataLoader(reader, num_epochs=2)
 
 
-@pytest.mark.parametrize('shuffling_queue_capacity', [20, 0])
+# shuffling_queue_capacity should be 0 or equal/greater to number of rows in file (50 in this test)
+@pytest.mark.parametrize('shuffling_queue_capacity', [50, 0])
 @pytest.mark.parametrize('reader_factory', [make_batch_reader, make_reader])
 @pytest.mark.parametrize('num_epochs', [1, 2, 3, None])
 def test_batched_data_loader_with_in_memory_cache(two_columns_non_petastorm_dataset,
@@ -235,7 +236,8 @@ def test_batched_data_loader_with_in_memory_cache(two_columns_non_petastorm_data
                                                   reader_factory,
                                                   num_epochs):
     batch_size = 10
-    extra_loader_params = dict(inmemory_cache_all=True, num_epochs=num_epochs)
+    extra_loader_params = dict(inmemory_cache_all=True, num_epochs=num_epochs,
+                               shuffling_queue_capacity=shuffling_queue_capacity)
     extra_reader_params = dict(num_epochs=1)
 
     with reader_factory(two_columns_non_petastorm_dataset.url,

